### DISCUSSION TITLE
API not fetch data with dynamic page

### DISCUSSION THREAD

**Discussion by Abahussain** [vercel/next.js#63866] [App Router]
The dynamic page is working but does not fetch the data from any API 

this is my dynamic page code 
please help me to fix this issue 

Path : src\app\Stocks\[ticker]\page.js

code :
"use client";
import { useRouter } from "next/navigation";
import { useEffect, useState } from "react";

const StockDetailsPage = () => {
  const router = useRouter();
  const [stockData, setStockData] = useState([]);

  useEffect(() => {
    fetchData();
  }, []);

  const fetchData = async () => {
    try {
      const apiUrl = "http://localhost:4000/stocks";
      const response = await fetch(apiUrl);
      if (!response.ok) {
        throw new Error("Failed to fetch data");
      }
      const data = await response.json();
      console.log("API Response:", data); // Log the API response
      if (!data || !data.stocks || !Array.isArray(data.stocks)) {
        throw new Error("Invalid data format");
      }
      setStockData(data.stocks); // Assuming the data is structured with a "stocks" array
    } catch (error) {
      console.error("Error fetching data:", error);
      setStockData([]);
    }
  };

  return (
    <>
      <div className="bg-background dark:bg-Background flex flex-col pl-[15px] pr-[15px] sm:pl-[100px] sm:pr-[40px] pt-2 sm:pt-4">
        <h1 className="mb-2">Stock Data</h1>
        <div className="rounded-2xl mb-4 sm:mb-8 dark:bg-gradient-to-tl from-[#101010] dark:to-[#161616] border dark:border-[#252525] bg-white flex-grow text-center">
          {stockData.map((stock) => (
            <div key={stock.ticker_id} className="p-4">
              <h2>{stock.ticker_id}</h2>
              <p>{stock.company_name}</p>
              <p>Price: {stock.price}</p>
              <p>Change: {stock.change}</p>
              <p>Change Percent: {stock.change_percent}</p>
              <p>Previous Close: {stock.prev_close}</p>
              <p>Open: {stock.open}</p>
              <p>Low: {stock.low}</p>
              <p>High: {stock.high}</p>
              <p>52 Weeks High: {stock["52_weeks_high"]}</p>
              <p>52 Weeks Low: {stock["52_weeks_low"]}</p>
              <p>Market Cap: {stock.market_cap}</p>
              <p>Volume: {stock.volume}</p>
              <p>Average Volume (3 Months): {stock.avg_vol_3_months}</p>
              <p>Turn Over: {stock.turn_over}</p>
              <p>Range: {stock.range}</p>
              <p>Shares Outstanding: {stock.shares_outstanding}</p>
              <p>Shares Float: {stock.shares_float}</p>
            </div>
          ))}
        </div>
      </div>
    </>
  );
};

export default StockDetailsPage;

I was trying with multiple APIs, working in the console only

**Comment by mutasim77** [comment]
It's really difficult to read your code because it's plain text, I would even say it's impossible. Next time when you ask a question, please provide the code _fully_ and _clearly_ within a **code block**. Also, write "what did you try and what kind of error it's throwing". So, anyway, I will give you an example with a dummy API, which should work without any issue:
```js
"use client";

import { useEffect, useState } from "react";

export default function UsersPage() {
    const [userData, setUserData] = useState(null);

    useEffect(() => {
        fetchData();
    }, [])

    const fetchData = async () => {
        try {
            const apiUrl = "https://reqres.in/api/users?page=2";
            const response = await fetch(apiUrl);

            if (!response.ok) {
                throw new Error("Failed to fetch data");
            }

            const data = await response.json();

            if (!data || !data.data || !Array.isArray(data.data)) {
                throw new Error("Invalid data format");
            }

            setUserData(data);
        } catch (error) {
            setUserData(null);
        }
    }

    return (
        <div>
            <h1>Users</h1>
            {userData && (
                <div>
                    {userData.data.map((user) => (
                        <div key={user.id}>
                            <p>ID: {user.id}</p>
                            <p>Email: {user.email}</p>
                            <p>First Name: {user.first_name}</p>
                            <p>Last Name: {user.last_name}</p>
                        </div>
                    ))}
                </div>
            )}
        </div>
    )
}
```

- Try this one; it should work. Also, don't forget to call this component : `<UsersPage/>`

- Also, I see your path is not correct. If it's not a typo, follow this structure: `app/stocks/[ticker]/page.js`
